### PR TITLE
fix(bazarr): change deployment method

### LINKED ISSUE #722: Current Bazarr install method deprecated
## Bug description
Swizzin currently installs Bazarr by [cloning the repository](https://github.com/swizzin/swizzin/blob/001bd702300e533ec7213911cc20428732239867/scripts/install/bazarr.sh#L32). As pointed out by @morpheus65535 in https://github.com/morpheus65535/bazarr/issues/1473#issuecomment-882467982 this is now deprecated, resulting in an error.

## Reproduction steps
1. Install package bazarr
2. Go to `/bazarr`
3. See error in bazarr logs

## Installed applications through swizzin
bazarr, deluge, delugeweb, nginx, panel, plex, prowlarr, radarr, sonarr

## Expected behavior
Swizzin should pull the latest asset release.

## Server Info
 - OS: Ubuntu 20.04
 - Arch: x86_64
 - Image Source: Installscript from Kimsufi / OVH

## Is Swizzin up to date?
HEAD is now at 001bd702300e533ec7213911cc20428732239867

## Additional context
See https://github.com/morpheus65535/bazarr/issues/1473

### PR DESCRIPTION
**Pull request by flying-sausages**
## Description

Git clone is no longer supported as an install method for bazarr
## Fixes issues: 
- Fixes #722

## Proposed Changes:
- style: separation into functions
- fix: #722
- fix: update nginx config
- fix: echos, logs, small logic, service

## Change Categories
- Bug fix 

## Checklist

- [x] Branch was made off the `develop` branch and the PR is targetting the `develop` branch
- [x] Docs have been made OR are not necessary
    - PR link: 
- [x] Changes to panel have been made OR are not necessary
    - PR link: 
- [x] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Code conforms to project structure [(See more)](https://swizzin.ltd/dev/structure)
- [x] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [x] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Testing was done
   - [ ] Tests created or no new tests necessary
   - [x] Tests executed

### Architectures

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|	Yup		|			|	Yup		|				|
| Bionic		|			|			|			|				|
| Buster		|			|			|			|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### CODE REVIEW

**Comment by flying-sausages** [comment]
@morpheus65535 I keep running into this issue with a timezone offset. I'm running on Ubuntu on Arm 20.04, which is a VM inside Parallels (sharing time or not does not impact this it seems). Timezone is correctly set, and `ntp` seems to have no issues with offsets. Running through `tzselect` also gives no movement on this. Any suggestions?

```
-- Logs begin at Tue 2021-02-16 22:49:28 CET, end at Mon 2021-07-19 19:06:28 CEST. --
Jul 19 19:06:27 m1 bazarr.test[16412]:   File "/opt/bazarr/bazarr/../libs/tzlocal/unix.py", line 165, in get_localzone
Jul 19 19:06:27 m1 bazarr.test[16412]:     _cache_tz = _get_localzone()
Jul 19 19:06:27 m1 bazarr.test[16412]:   File "/opt/bazarr/bazarr/../libs/tzlocal/unix.py", line 90, in _get_localzone
Jul 19 19:06:27 m1 bazarr.test[16412]:     utils.assert_tz_offset(tz)
Jul 19 19:06:27 m1 bazarr.test[16412]:   File "/opt/bazarr/bazarr/../libs/tzlocal/utils.py", line 46, in assert_tz_offset
Jul 19 19:06:27 m1 bazarr.test[16412]:     raise ValueError(msg)
Jul 19 19:06:27 m1 bazarr.test[16412]: ValueError: Timezone offset does not match system offset: 0 != 7200. Please, check your config files.
Jul 19 19:06:28 m1 bazarr.test[16372]: Bazarr starting...
Jul 19 19:06:28 m1 bazarr.test[16372]: Bazarr exited.
Jul 19 19:06:28 m1 systemd[1]: bazarr.service: Succeeded.
```

**Comment by morpheus65535** [comment]
@flying-sausages it's a known bug with some timezones: https://github.com/regebro/tzlocal/issues/79

**Comment by flying-sausages** [comment]
> @flying-sausages it's a known bug with some timezones: [regebro/tzlocal#79](https://github.com/regebro/tzlocal/issues/79)

Thanks, that helped me work around it. Do I take it that if the install worked previously, this issue should not arise with this new release?

**Comment by morpheus65535** [comment]
> > @flying-sausages it's a known bug with some timezones: [regebro/tzlocal#79](https://github.com/regebro/tzlocal/issues/79)
> 
> Thanks, that helped me work around it. Do I take it that if the install worked previously, this issue should not arise with this new release?

Yes it will work. Nothing new here.

**Comment by flying-sausages** [comment]
Cool, thanks!

This is good to go.

File: scripts/update/bazarr.sh [hunk 1]
```diff
@@ -29,4 +29,42 @@ if [[ -f /install/.bazarr.lock ]]; then
...
+        }
+        rm /tmp/bazarr.zip
+
+        sudo -u "${user}" bash -c "/opt/.venv/bazarr/bin/pip3 install -r /opt/bazarr/requirements.txt" >> $log 2>&1 || {
+            echo_error "Dependencies failed to install"
+            exit 1
+        }
+
+        cp -R /tmp/bazarr_data /opt/bazarr/data
+        chown -R "${user}": /opt/bazarrgit
```
Comment (she11sh0cked) [inline scripts/update/bazarr.sh:63]: Should be `/opt/bazarr`, right?

**Review by she11sh0cked — COMMENTED** [review]
Ortherwise it seems fine.

File: scripts/update/bazarr.sh (hunk 1 above)
Comment (flying-sausages) [reply scripts/update/bazarr.sh:63]: fml yeah cheers

**Review by she11sh0cked — APPROVED** [review]

**Comment by liaralabs** [comment]
I have never had good luck with the internal dependency updater and as further proof, try this:

```
user=<updatemefortests>
rm -rf /opt/.venv/bazarr
python3 -m venv /opt/.venv/bazarr
chown -R ${user}: /opt/.venv/bazarr
```

Then try running bazarr. Bazarr will never actually successfully use its internal updater when placed in a venv

```
Bazarr is restarting...
2021-07-26 23:47:23,226 - root                             (7fd918c48740) :  INFO (init:58) - BAZARR installing requirements...
2021-07-26 23:47:25,927 - root                             (7fd918c48740) :  INFO (init:68) - BAZARR requirements installed.
2021-07-26 23:47:25,928 - root                             (7fd918c48740) :  INFO (init:75) - Bazarr is being restarted...
Bazarr is restarting...
2021-07-26 23:47:27,067 - root                             (7fc9f89a3740) :  INFO (init:58) - BAZARR installing requirements...
2021-07-26 23:47:29,784 - root                             (7fc9f89a3740) :  INFO (init:68) - BAZARR requirements installed.
2021-07-26 23:47:29,784 - root                             (7fc9f89a3740) :  INFO (init:75) - Bazarr is being restarted...
```

This will just loop forever if you let it.

So, while we are here we should implement a work around to the internal update mechanism as requirements do appear to change from time to time. Thus I would recommend adding this to the systemd service:

```
ExecStartPre=/opt/.venv/bazarr/bin/pip install -r /opt/bazarr/requirements.txt
```

Otherwise, other than the TZ issue which is likely just due to virtual environments it looks good to me

**Comment by morpheus65535** [comment]
@liaralabs what version of virtualenv are you using? Old one seems to have bugs when returning sys.executable. Upgrade it and you should be fine.

**Comment by liaralabs** [comment]
@morpheus65535 I do all my testing against Debian 10 at the moment and I use `python3 -m venv` not `virtualenv` to create venvs. It looks like the venv module itself is a python built-in module and can't be upgraded without deploying a completely new version of python (i.e. through pyenv). That said, we've been defaulting to 3.7.7 for pyenv installs as well.

That said, the python compile time can be lengthy and we would generally hope to avoid deploying a python environment when the one shipped by the distro is already sufficient enough

However, giving it a test under focal provides an actual error message with more details:

```
 ERROR (init:65) - BAZARR requirements.txt installation result: b"ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
 ```

It then quits after this message, as opposed to the loop on Debian 10

fwiw I receive the same error with venv or virtualenv to create the venv.

**Comment by morpheus65535** [comment]
@liaralabs so it seems that --user cannot be used with pip under virtualenv. Make sense. If you edit the command in bazarr/init.py to remove it, does it works? If it does, I can try to detect if virtualenv is activated and don't use --user in that case.

**Comment by liaralabs** [comment]
@morpheus65535 Yes, it looks like removing the `--user` argument from the `pip` command indeed allows the internal dependency updater to function correctly -- even on Debian 10, which was stuck in a loop without the explicit error message, this has allowed the internal dep updater to continue.

Some form of venv detection would be great!

It appears to be starting fine on Bullseye with the same tweak (despite the python3.9 warning, it appears to be running)

**Comment by morpheus65535** [comment]
@liaralabs thanks for testing! I'll add the detection in upcoming 0.9.7-beta.4 release and you should be able to test it out.

**Comment by flying-sausages** [comment]
0.9.7 was released a couple days ago, @liaralabs could you please try again?

**Comment by liaralabs** [comment]
Only thing I bumped into during testing is that `unzip` is not installed though it is a dependency for unpacking the package. On a totally fresh, minimal test system with no installed packages other than nginx, the installer appears to have failed due to this.

Feeling like maybe we should make unzip a global depend though; however packages that require it up to this point (i.e. autodl, nextcloud, nzbhydra) usually just install it themselves.

Looks good to me and the PR works with unzip installed.

**Comment by flying-sausages** [comment]
> Feeling like maybe we should make unzip a global depend though

https://github.com/swizzin/swizzin/pull/736 gotcha

**Pull request merged**